### PR TITLE
Sebs Workings

### PR DESCRIPTION
**Pull request by PrenSJ2**
﻿

### CODE REVIEW

**Comment by PrenSJ2** [comment]
image

File: app/pipedrive/_schema.py [hunk 1]
```diff
@@ -224,23 +226,34 @@ class PDDeal(PipedriveBaseModel):
     status: str
     obj_type: Literal['deal'] = Field('deal', exclude=True)
 
+    # # These are all custom fields
```
Comment (PrenSJ2) [inline app/pipedrive/_schema.py:229]: need to add this stuff back in otherwise hermes creates two pipedrive deals

**Comment by PrenSJ2** [comment]
image
Edit / create an admin, pipedrive_id shouldn't be required

File: app/callbooker/_booking.py [hunk 2]
```diff
@@ -50,7 +50,7 @@ async def create_meeting_gcal_event(meeting: Meeting):
     g_cal = AdminGoogleCalendar(admin_email=admin.email)
     g_cal.create_cal_event(
         description=meeting_template['description'].format(**meeting_templ_vars),
-        summary=meeting_template['summary'].format(**meeting_templ_vars),
```
Comment (tomhamiltonstubber) [inline app/callbooker/_booking.py:53]: Is `meeting_template['summary']` still there? Does it do anything?

File: app/callbooker/_process.py [hunk 3]
```diff
@@ -94,14 +94,15 @@ async def get_or_create_deal(company: Company, contact: Contact) -> Deal:
     config = await get_config()
     if not deal:
         match company.price_plan:
-            case Company.PP_PAYG:
+            case plan if Company.PP_PAYG in plan:
```
Comment (tomhamiltonstubber) [inline app/callbooker/_process.py:97]: If `price_plan` is being validated from the tc2 webhook then better to use pydantic validator to remove the `-`

File: app/hermes/views.py [hunk 4]
```diff
@@ -32,7 +32,9 @@ async def choose_sales_person(plan: str) -> Admin.pydantic_schema():
     else:
         next_sales_person = admin_ids[0]
     schema = Admin.pydantic_schema()
-    return await schema.from_tortoise_orm(admins[next_sales_person])
+
+    returned_admin = await schema.from_tortoise_orm(admins[next_sales_person])
+    return returned_admin
```
Comment (tomhamiltonstubber) [inline app/hermes/views.py:37]: reverse

File: app/pipedrive/_schema.py [hunk 5]
```diff
@@ -136,26 +138,29 @@ async def company_dict(self) -> dict:
 class Person(PipedriveBaseModel):
     id: Optional[int] = Field(None, exclude=True)
     name: str
-    primary_email: Optional[str] = ''
+    primary_email: Optional[str] = Field('', custom=True)
     phone: Optional[str] = ''
-    address_country: Optional[str] = None
+    address_country: Optional[str] = Field(None, custom=True)
```
Comment (tomhamiltonstubber) [inline app/pipedrive/_schema.py:143]: I don't think either of these are custom fields? Especially email?

File: app/pipedrive/_schema.py [hunk 6]
```diff
@@ -222,6 +227,12 @@ class PDDeal(PipedriveBaseModel):
     pipeline_id: fk_field(Pipeline, 'pd_pipeline_id')
     stage_id: fk_field(Stage, 'pd_stage_id')
     status: str
+
+    # These are all custom fields
+    hermes_deal_id: fk_field(Deal, 'id', null_if_invalid=True) = Field('', custom=True)
```
Comment (tomhamiltonstubber) [inline app/pipedrive/_schema.py:232]: Change these all to just `hermes_id`

File: app/pipedrive/views.py [hunk 7]
```diff
@@ -41,4 +43,14 @@ async def callback(event: PipedriveEvent, tasks: BackgroundTasks):
         if company and company.tc2_agency_id:
             # We only update the client if the deal has a company with a tc2_agency_id
             tasks.add_task(update_client_from_company, company)
+    elif event.meta.object == 'activity':
+        pass
```
Comment (tomhamiltonstubber) [inline app/pipedrive/views.py:47]: If we don't want to accept webhooks for Activites, can we not just select that option in the Pipedrive webhook setup?

File: app/pipedrive/views.py [hunk 8]
```diff
@@ -41,4 +43,14 @@ async def callback(event: PipedriveEvent, tasks: BackgroundTasks):
...
     return {'status': 'ok'}
+
+
+@pipedrive_router.post('/callback/debug/', name='Pipedrive debug callback')
+async def debug_callback(body: Request):
+    data = await body.json()
+    app_logger.info(f'Debug: event received for {data}')
+    r = requests.post('http://localhost:8000/pipedrive/callback/', json=data)
+    return r.json
```
Comment (tomhamiltonstubber) [inline app/pipedrive/views.py:56]: Remove

File: app/settings.py [hunk 9]
```diff
@@ -54,6 +54,12 @@ class Settings(BaseSettings):
         'https://www.googleapis.com/robot/v1/metadata/x509/tc-hubspot%40tc-hubspot-314214.iam.gserviceaccount.com'
     )
 
+    # Testing
+    # Admin user to create when testing locally
+    CREATE_TESTING_ADMIN: bool = False
+    tc2_admin_id: int = 66
+    pd_owner_id: int = 15708604
```
Comment (tomhamiltonstubber) [inline app/settings.py:61]: Definitely remove

File: app/tc2/_process.py [hunk 10]
```diff
@@ -47,14 +48,15 @@ async def _get_or_create_deal(company: Company, contact: Contact | None) -> Deal
     config = await get_config()
     if not deal:
         match company.price_plan:
-            case Company.PP_PAYG:
+            case plan if Company.PP_PAYG in plan:
```
Comment (tomhamiltonstubber) [inline app/tc2/_process.py:51]: As above

File: app/tc2/_schema.py [hunk 11]
```diff
@@ -38,6 +38,7 @@ class _TCAgency(HermesBaseModel):
     status: str
     paid_invoice_count: int
     created: datetime = Field(exclude=True)
+    # price_plan: str
```
Comment (tomhamiltonstubber) [inline app/tc2/_schema.py:41]: ?

File: app/tc2/_schema.py [hunk 12]
```diff
@@ -88,10 +90,12 @@ def company_dict(self):
             tc2_status=self.meta_agency.status,
             name=self.meta_agency.name,
             country=self.meta_agency.country,
+            website=self.meta_agency.website,
             support_person=self.support_person,  # noqa: F821 - Added in validation
             sales_person=self.sales_person,  # noqa: F821 - Added in validation
             bdr_person=self.bdr_person,  # noqa: F821 - Added in validation
             paid_invoice_count=self.meta_agency.paid_invoice_count,
+            # price_plan=self.meta_agency.price_plan,
```
Comment (tomhamiltonstubber) [inline app/tc2/_schema.py:98]: ?

File: app/tc2/api.py [hunk 13]
```diff
@@ -7,8 +7,9 @@
 
 
 async def tc2_request(url: str, *, method: str = 'GET', data: dict = None) -> dict:
-    headers = {'Authorization': f'token {settings.tc2_api_key}'}
-    r = session.request(method=method, url=f'{settings.tc2_base_url}/api/{url}', data=data, headers=headers)
+    api_key = settings.tc2_api_key.decode('utf-8')
```
Comment (tomhamiltonstubber) [inline app/tc2/api.py:10]: Why is it encoded? Shouldn't be.

File: app/tc2/views.py [hunk 14]
```diff
@@ -38,3 +38,9 @@ async def callback(
         if company:
             tasks.add_task(pd_post_process_client_event, company, deal)
     return {'status': 'ok'}
+
+
+@tc2_router.post('/debug_log/', name='debug log')
+async def debug_log(event_data: dict):
+    app_logger.info("Captured event data for debugging:")
+    return {'status': 'ok'}
```
Comment (tomhamiltonstubber) [inline app/tc2/views.py:46]: Remove

File: app/utils.py [hunk 15]
```diff
@@ -40,17 +40,24 @@ async def get_config() -> 'Config':
...
-    #     await Admin.create(
-    #         email='testing@tutorcruncher.com',
-    #         username='testing@tutorcruncher.com',
-    #         password='testing',
-    #         is_bdr_person=True,
-    #         is_sales_person=True,
-    #         tc2_admin_id=66,
-    #         pd_owner_id=15396545,
-    #     )
+    # if CREATE_TESTING_ADMIN env var is true
```
Comment (tomhamiltonstubber) [inline app/utils.py:43]: Don't need this line

File: app/utils.py [hunk 16]
```diff
@@ -40,17 +40,24 @@ async def get_config() -> 'Config':
...
+                is_bdr_person=True,
+                is_sales_person=True,
+                tc2_admin_id=settings.tc2_admin_id,
+                pd_owner_id=settings.pd_owner_id,
+            )
+        if not await Pipeline.exists():
+            await Pipeline.create(
+                name='pipeline',
+                pd_pipeline_id=1,
+            )
```
Comment (tomhamiltonstubber) [inline app/utils.py:61]: Is this needed? Think not.

File: tests/test_callbooker.py [hunk 17]
```diff
@@ -697,10 +697,12 @@ async def test_generate_support_link(self):
         company = await Company.create(
             name='Junes Ltd', website='https://junes.com', country='GB', tc2_cligency_id=10, sales_person=admin
         )
+        api_key = settings.tc2_api_key.decode('utf-8')
```
Comment (tomhamiltonstubber) [inline tests/test_callbooker.py:700]: As above, this shouldn't be a bytesstring

**Review by tomhamiltonstubber — CHANGES_REQUESTED** [review]

**Comment by PrenSJ2** [comment]
And Re Read through Readme and give more context and description as if you have no idea what this is :face_with_thermometer:

File: app/pipedrive/_schema.py (hunk 5 above)
Comment (PrenSJ2) [reply app/pipedrive/_schema.py:143]: you are correct on email, we need `address_country` tho

File: app/models.py [hunk 18]
```diff
@@ -184,12 +191,24 @@ class Company(models.Model):
...
+    def _price_plan(cls, v):
+        # Extract the part after the hyphen
+        plan = v.split('-')[1] if '-' in v else v
+
+        # Validate the extracted part
+        valid_plans = (cls.PP_PAYG, cls.PP_STARTUP, cls.PP_ENTERPRISE)
+        if plan not in valid_plans:
+            raise ValueError(f"Invalid price plan: {v}")
+
+        return plan
```
Comment (tomhamiltonstubber) [inline app/models.py:204]: This is wrong, which you would know if you had a test

File: app/pipedrive/_schema.py [hunk 19]
```diff
@@ -90,7 +90,8 @@ class Config:
 class Organisation(PipedriveBaseModel):
     id: Optional[int] = Field(None, exclude=True)
     name: str
-    address_country: Optional[str] = None
+    # address_country: Optional[str] = None
```
Comment (tomhamiltonstubber) [inline app/pipedrive/_schema.py:93]: Commented out but not deleted?

File: app/pipedrive/_schema.py [hunk 20]
```diff
@@ -136,26 +139,33 @@ async def company_dict(self) -> dict:
...
     name: str
-    primary_email: Optional[str] = ''
+    email: Optional[str] = ''
     phone: Optional[str] = ''
-    address_country: Optional[str] = None
     owner_id: Optional[fk_field(Admin, 'pd_owner_id')] = None
     org_id: Optional[fk_field(Company, 'pd_org_id', null_if_invalid=True)] = None
 
+    # These are all custom fields
+    address_country: Optional[str] = Field(None, custom=True)
```
Comment (tomhamiltonstubber) [inline app/pipedrive/_schema.py:148]: Why is this a custom field? Surely it should follow the same logic as Orgs?

File: app/pipedrive/_schema.py [hunk 21]
```diff
@@ -222,6 +232,12 @@ class PDDeal(PipedriveBaseModel):
     pipeline_id: fk_field(Pipeline, 'pd_pipeline_id')
     stage_id: fk_field(Stage, 'pd_stage_id')
     status: str
+
+    # These are all custom fields
+    hermes_deal_id: fk_field(Deal, 'id', null_if_invalid=True) = Field('', custom=True)
```
Comment (tomhamiltonstubber) [inline app/pipedrive/_schema.py:237]: This isn't done?

File: app/pipedrive/views.py (hunk 7 above)
Comment (tomhamiltonstubber) [reply app/pipedrive/views.py:47]: This isn't done?

**Review by tomhamiltonstubber — CHANGES_REQUESTED** [review]

File: app/pipedrive/_schema.py (hunk 20 above)
Comment (PrenSJ2) [reply app/pipedrive/_schema.py:148]: Unfortunately not, there is no address field on a PD Person

Comment (tomhamiltonstubber) [reply app/pipedrive/_schema.py:147]: Remove then

File: app/admin/resources.py [hunk 22]
```diff
@@ -172,7 +172,7 @@ class DealResource(Model):
         Field('pipeline_id', input_=inputs.ForeignKey(model=Pipeline)),
         Field('stage_id', input_=inputs.ForeignKey(model=Stage)),
         Field('company_id', input_=inputs.ForeignKey(model=Company)),
-        Field('contact_id', input_=inputs.ForeignKey(model=Contact)),
+        Field('hermes_id', input_=inputs.ForeignKey(model=Contact)),
```
Comment (tomhamiltonstubber) [inline app/admin/resources.py:175]: Fix this

File: app/pipedrive/_schema.py [hunk 23]
```diff
@@ -111,13 +112,14 @@ async def from_company(cls, company: Company) -> 'Organisation':
             **_remove_nulls(
                 name=company.name,
                 owner_id=(await company.sales_person).pd_owner_id,
-                address_country=company.country,
```
Comment (tomhamiltonstubber) [inline app/pipedrive/_schema.py:114]: See if you can use `address_country` or possible `country_code`

File: app/pipedrive/_schema.py [hunk 24]
```diff
@@ -136,28 +138,35 @@ async def company_dict(self) -> dict:
 class Person(PipedriveBaseModel):
     id: Optional[int] = Field(None, exclude=True)
     name: str
-    primary_email: Optional[str] = ''
+    email: Optional[str] = ''
```
Comment (tomhamiltonstubber) [inline app/pipedrive/_schema.py:141]: Check webhook to make sure it's not primary email

Comment (tomhamiltonstubber) [reply app/pipedrive/_schema.py:141]: We think this is wrong :space_invader:

File: app/pipedrive/_schema.py [hunk 25]
```diff
@@ -136,28 +138,35 @@ async def company_dict(self) -> dict:
...
             phone=contact.phone,
             address_country=contact.country,
             org_id=company.pd_org_id,
+            hermes_id=contact.id,
         )
+        obj = await cls.set_custom_field_vals(obj)
+        return obj
 
-    @validator('phone', pre=True)
+    @validator('phone', 'email', pre=True)
```
Comment (tomhamiltonstubber) [inline app/pipedrive/_schema.py:169]: remove change
```suggestion
    @validator('phone', pre=True)
```

File: app/settings.py [hunk 26]
```diff
@@ -54,6 +54,9 @@ class Settings(BaseSettings):
         'https://www.googleapis.com/robot/v1/metadata/x509/tc-hubspot%40tc-hubspot-314214.iam.gserviceaccount.com'
     )
 
+    # Testing
+    TESTING: bool = False
```
Comment (tomhamiltonstubber) [inline app/settings.py:58]: _small_

File: app/tc2/_schema.py [hunk 27]
```diff
@@ -62,6 +75,7 @@ class TCClient(HermesBaseModel):
     meta_agency: _TCAgency = Field(exclude=True)
     user: _TCUser
     status: str
+    website: Optional[str] = None
```
Comment (tomhamiltonstubber) [inline app/tc2/_schema.py:78]: ?

File: app/utils.py [hunk 28]
```diff
@@ -40,17 +40,20 @@ async def get_config() -> 'Config':
...
-    #         username='testing@tutorcruncher.com',
-    #         password='testing',
-    #         is_bdr_person=True,
-    #         is_sales_person=True,
-    #         tc2_admin_id=66,
-    #         pd_owner_id=15396545,
-    #     )
+    from app.models import Admin, Pipeline
+
+    if not settings.TESTING and await Admin.filter(username='testing@tutorcruncher.com').exists():
```
Comment (tomhamiltonstubber) [inline app/utils.py:45]:
```suggestion
    admins_exist = await Admin.exists()
    if not admins_exist and not settings.testing and settings.dev_mode:
```

File: app/utils.py [hunk 29]
```diff
@@ -40,17 +40,20 @@ async def get_config() -> 'Config':
...
+            username='testing@tutorcruncher.com',
+            password='testing',
+            is_bdr_person=True,
+            is_sales_person=True,
+        )
+    if not await Pipeline.exists():
+        await Pipeline.create(
+            name='pipeline',
+            pd_pipeline_id=1,
+        )
```
Comment (tomhamiltonstubber) [inline app/utils.py:57]: Don't need this, please delete

File: tests/test_admin.py [hunk 30]
```diff
@@ -6,6 +7,7 @@
 class AdminTestCase(HermesTestCase):
     async def asyncSetUp(self) -> None:
         await super().asyncSetUp()
+        settings.TESTING = True
```
Comment (tomhamiltonstubber) [inline tests/test_admin.py:10]: Do this in the `super()` method

**Review by tomhamiltonstubber — CHANGES_REQUESTED** [review]

File: app/settings.py [hunk 31]
```diff
@@ -55,7 +57,7 @@ class Settings(BaseSettings):
     )
 
     # Testing
-    TESTING: bool = False
```
Comment (tomhamiltonstubber) [inline app/settings.py]: ?

File: app/tc2/_schema.py [hunk 32]
```diff
@@ -43,12 +43,10 @@ class _TCAgency(HermesBaseModel):
     @validator('price_plan')
     def _price_plan(cls, v):
         # Extract the part after the hyphen
-        plan = v.split('-')[1] if '-' in v else v
-
-        # Validate the extracted part
-        valid_plans = (Company.PP_PAYG, Company.PP_STARTUP, Company.PP_ENTERPRISE)
```
Comment (tomhamiltonstubber) [inline app/tc2/_schema.py]: What did you remove this check?

File: app/utils.py [hunk 33]
```diff
@@ -40,20 +40,18 @@ async def get_config() -> 'Config':
         config = await Config.create()
 
     # When testing locally, you can add your own admin user here. Set the pd_owner_id to your own Pipedrive user ID.
-    from app.models import Admin, Pipeline
+    from app.models import Admin
 
-    if not settings.TESTING and await Admin.filter(username='testing@tutorcruncher.com').exists():
+    admin_exists = await Admin.exists()
+    debug(settings.dev_mode, admin_exists, settings.testing)
```
Comment (tomhamiltonstubber) [inline app/utils.py:46]: !

**Review by tomhamiltonstubber — CHANGES_REQUESTED** [review]

**Pull request merged**